Overhaul docs (#468)
nautobot/nautobot-app-ssot: docs/dev/jobs.md

# Tutorial: Developing a Data Source Integration

This tutorial will walk you through building a custom data source (i.e. synchronizing data _to_ Nautobot) SSoT integration with a remote system. We will be using static data as the remote system, but it should be easy enough to substitute this later for any external systemy ou want to integrate with.

*Familiarity with [DiffSync](https://diffsync.readthedocs.io/en/latest/) and with developing [Nautobot Jobs](https://nautobot.readthedocs.io/en/latest/additional-features/jobs/) is a plus, but important concepts will be explained or linked to along the way.**

## Creating a New Nautobot App

To start building your own SSoT integration, you first need a project as well as the accompanying development environment. Network to Code provides the [Cookiecutter Nautobot App](https://github.com/nautobot/cookiecutter-nautobot-app) to make this as easy as possible for you. Check out the README of that project and bake a `nautobot-app` cookie.

!!! note

>     The `nautobot-app-ssot` cookie is not yet updated to support the newest paradigms of building SSoT integrations, which is why it is recommended to instead use the `nautobot-app` cookie.

Author: Tracking this [here](https://github.com/nautobot/cookiecutter-nautobot-app/issues/202) and want to do that first, as it also addresses another comment of yours.

### Building the Local adapter

First, we define the adapter to load data from the local Nautobot instance - this is very straight-forward.

```python
# tutorial_ssot_app/ssot/adapters.py
from nautobot_ssot.contrib import NautobotAdapter
from tutorial_ssot_app.ssot.models import VLANModel

class MySSoTNautobotAdapter(NautobotAdapter):
    vlan = VLANModel  # Map your model into the adapter
    top_level = ("vlan",)  # Tell the adapter to consider your model
```

Now create a couple of VLANs in the web GUI of Nautobot so that the adapter has some data to load. Once you're done, you can try out the adapter as follows:

```python
from tutorial_ssot_app.ssot.adapters import MySSoTNautobotAdapter

adapter = MySSoTNautobotAdapter(job=None)
adapter.load()
adapter.count("vlan")  # This will return the amount of VLANs that were loaded from Nautobot

test_vlan = adapter.get_all("vlan")[0]  # Now we retrieve an example VLAN to test the model with
test_vlan.update(attrs={"description": "My updated description"})  # Verify the update has worked using the web GUI
test_vlan.delete()  # Verify the deletion has worked using the web GUI
```

### Building the Remote Adapter

> To synchronize a diff, we need to pull data from both sides. Therefore, we now need to build the remote adapter. In this example we are reading static data, but feel free to substitute this adapter with your remote system of choice.

Author: I wouldn't consider Nautobot a system of record, only the remote side. What do you think? I have updated this as such.

!!! note
    Note that we are not subclassing from `nautobot_ssot.contrib.NautobotAdapter` here as we don't want the SSoT framework to auto-derive the loading implementation of the adapter for us, but rather want to handle this ourselves.

```python
# tutorial_ssot_app/ssot/adapters.py
from diffsync import DiffSync

from tutorial_ssot_app.ssot.models import VLANModel

class MySSoTRemoteAdapter(DiffSync):
    """DiffSync adapter for remote system."""
    vlan = VLANModel
    top_level = ("vlan",)

    def __init__(self, *args, api_client, job=None, **kwargs):
        super().__init__(*args, **kwargs)
        self.api_client = api_client
        self.job = job

    def load(self):
        vlans = [
            {"name": "Servers", "vid": 100, "description": "Server VLAN Datacenter", "status__name": "Active"},
            {"name": "Printers", "vid": 200, "description": "Printer VLAN Office", "status__name": "Deprecated"},
            {"name": "Clients", "vid": 300, "description": "Client VLAN Office", "status__name": "Active"},
        ]
        for vlan in vlans:

>             loaded_vlan = self.vlan(

Author: I want to keep this as simple as possible. I would include this in a more thorough reference section possibly. What do you think?

                name=vlan["name"],
                vid=vlan["vid"],
                description=vlan["description"],
                status__name=vlan["status__name"]
            )
            self.add(loaded_vlan)
```

Now you can verify that this adapter is working in a similar manner to how we did with the Nautobot adapter, feel free to try this!

## Putting it Together in a Job

> Having built the model and both adapters, we can now put everything together into a job to finish up the integration.

Reviewer: ```suggestion
Having built the model and both Adapters, we can now put everything together into a Job to finish up the integration.
```

```python
# tutorial_ssot_app/jobs.py
from nautobot_ssot.jobs.base import DataSource
from nautobot.apps.jobs import Job, register_jobs
from tutorial_ssot_app.ssot.adapters import MySSoTRemoteAdapter, MySSoTNautobotAdapter

> class TutorialDataSource(DataSource, Job):

Reviewer: ```suggestion
class TutorialDataSource(DataSource):
```
We no longer need to inherit from Job in SSoT 2.x+.

    class Meta:
        name = "Example Data Source"

    def load_source_adapter(self):
        self.source_adapter = MySSoTRemoteAdapter(api_client=APIClient(), job=self)
        self.source_adapter.load()

    def load_target_adapter(self):
        self.target_adapter = MySSoTNautobotAdapter(job=self)
        self.target_adapter.load()

register_jobs(TutorialDataSource)
```

> At this point, the job will show up in the web GUI, and you can enable it and even run it! You should see the objects you specified in the remote adapter being synchronized into Nautobot now. Also take a look at the "SSoT Sync Details" button in the top right of the job result page to get some more information on what diffsync is doing.

Reviewer: ```suggestion
At this point, the Job will show up in the web GUI, at which point you can enable it and even run it! You should see the objects you specified in the remote Adapter being synchronized into Nautobot now. In addition, clicking on the "SSoT Sync Details" button in the top right of the Job result page provides additional information on what DiffSync is doing during the synchronization process.
```

> The above example shows the simplest field type (an attribute on the model), however, to build a production implementation you will need to understand how to identify different variants of fields by following the [modeling docs](../dev/modeling.md).

Reviewer: ```suggestion
The above example shows the simplest field type (an attribute on the model), however, to build a production implementation you will need to understand how to identify different variants of fields by following the [Modeling docs](../dev/modeling.md).
```
